### PR TITLE
Report filter selectivity for pushed down filters

### PR DESCRIPTION
**Pull request by mbasmanova**
- Add `ConnectorPageSource#getCompletedPositions` SPI to fetch number of rows processed by the connector and use this value in TableScanOperator and ScanFilterProjectOperator to populate `OperatorStats#rawInputPositions`.
- Update PlanPrinter to report filter selectivity for TableScanNode.

```
select orderkey from lineitem where orderkey % 2 = 0

     - TableScan[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=lineitem, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.lineitem{filter=((orderkey) % (CAST(VARCHAR 2 AS bigint))) = (CAST(VARCHAR 0 AS bigint))}]'}, grouped = false] => [orderkey:bigint] 
             CPU: 32.00ms (100.00%), Scheduled: 39.00ms (100.00%), Output: 30050 rows (264.11kB)                                                                                                                                                                                                                           
             Input avg.: 15025.00 rows, Input std.dev.: 0.65%                                                                                                                                                                                                                                                              
             LAYOUT: tpch.lineitem{filter=((orderkey) % (CAST(VARCHAR 2 AS bigint))) = (CAST(VARCHAR 0 AS bigint))}                                                                                                                                                                                                        
             orderkey := orderkey:bigint:0:REGULAR                                                                                                                                                                                                                                                                         
             Input: 30050 rows (264.11kB), Filtered: 0.00%                                                                                                                                                                                                                                                                 
             Raw input: 60175 rows (1.38MB), Filtered: 50.06%                                                                                                                                                                                                                                                              


select orderkey from lineitem where linenumber < 2

     - TableScan[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=lineitem, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.lineitem{domains={linenumber=[ [(<min>, 2)] ]}}]'}, grouped = false] => [orderkey:bigint] 
             CPU: 29.00ms (100.00%), Scheduled: 90.00ms (100.00%), Output: 15000 rows (131.83kB)                                                                                                                                                                                 
             Input avg.: 7500.00 rows, Input std.dev.: 0.16%                                                                                                                                                                                                                     
             LAYOUT: tpch.lineitem{domains={linenumber=[ [(<min>, 2)] ]}}                                                                                                                                                                                                        
             orderkey := orderkey:bigint:0:REGULAR                                                                                                                                                                                                                               
             Input: 15000 rows (131.83kB), Filtered: 0.00%                                                                                                                                                                                                                       
             Raw input: 60175 rows (1.38MB), Filtered: 75.07%  


select orderkey from lineitem

     - TableScan[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=lineitem, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.lineitem{}]'}, grouped = false] => [orderkey:bigint] 
             CPU: 16.00ms (100.00%), Scheduled: 18.00ms (100.00%), Output: 60175 rows (528.88kB)                                                                                                                                            
             Input avg.: 30087.50 rows, Input std.dev.: 0.50%                                                                                                                                                                               
             LAYOUT: tpch.lineitem{}                                                                                                                                                                                                        
             orderkey := orderkey:bigint:0:REGULAR                                                                                                                                                                                          
             Input: 60175 rows (528.88kB), Filtered: 0.00% 


select orderkey + 1 from lineitem where linenumber < 2

     - ScanProject[table = TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=lineitem, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.lineitem{domains={linenumber=[ [(<min>, 2)] ]}}]'}, grouped = false] => [expr:bigint] 
             CPU: 20.00ms (100.00%), Scheduled: 23.00ms (100.00%), Output: 15000 rows (131.83kB)                                                                                                                                                                                       
             Input avg.: 7500.00 rows, Input std.dev.: 0.16%                                                                                                                                                                                                                           
             expr := (orderkey) + (CAST(VARCHAR 1 AS bigint))                                                                                                                                                                                                                          
             LAYOUT: tpch.lineitem{domains={linenumber=[ [(<min>, 2)] ]}}                                                                                                                                                                                                              
             orderkey := orderkey:bigint:0:REGULAR                                                                                                                                                                                                                                     
             Input: 15000 rows (131.83kB), Filtered: 0.00%                                                                                                                                                                                                                             
             Raw input: 60175 rows (1.38MB), Filtered: 75.07%   
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java [hunk 1]
```diff
@@ -254,10 +255,15 @@ public Page getOutput()
 
             // update operator stats
             long endCompletedBytes = source.getCompletedBytes();
+            long endCompletedPositions = source.getCompletedPositions();
+            if (endCompletedPositions < 0) {
+                endCompletedPositions = completedPositions + page.getPositionCount();
```
Comment (sayhar) [inline presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: Trying to understand -- why would statement ever be true?

Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: Connectors that don't implement the new method will use the default implementation that returns -1.

File: presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java [hunk 2]
```diff
@@ -254,10 +255,15 @@ public Page getOutput()
 
             // update operator stats
             long endCompletedBytes = source.getCompletedBytes();
+            long endCompletedPositions = source.getCompletedPositions();
+            if (endCompletedPositions < 0) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: This logic should go inside each connector. Check the other comment.

**Comment by mbasmanova** [comment]
@highker 

> recordRawInputWithTiming is the stats reported by each connector. The stats could be incorrect. On the other hand recordProcessedInput is the precise stats we will need to report. Having it relying on connectors will be a problem. Why can't we just use position count from the output page?

Before filter pushdown, ScanFilterProject operator reported the original row count (input positions) and filtered row count (output positions). This information is used to report filter selectivity in the output of EXPLAIN command and is available in the operator stats. With filter pushdown, we now have just TableScan operator which doesn't know the original row count as it receives already filtered rows from the connector. Hence, we are missing the filter selectivity stats.

One option could be to add a new field rowInputPositions to OperatorContext and OperatorStats and extend recordRawInputWithTiming to include row count (currently it only reports size in bytes). What do you think?

> Instead of having the default, we should make it true abstract and force the connectors to implement this interface. 

I'll take a look at how to do that.

**Comment by mbasmanova** [comment]
@highker 

> One option could be to add a new field rowInputPositions to OperatorContext and OperatorStats and extend recordRawInputWithTiming to include row count (currently it only reports size in bytes). 

James, I made this change and update PR description to show the output of the `explain analyze`. Would you take another look?

**Comment by oerling** [comment]
This will certainly answer all immediate needs. 

 

James, as a thing to keep in mind, there are expected use cases for more general mergeable metadata about operator performance. This should in no way affect this PR though. One expected case is tracking column level selectivities as an indication for physical optimization. So, we’ll want to evolve this in the direction of an interface to operator performance metadata that can encapsulate raw read positions but also filter stats. In the case of federation, this can encapsulate whatever stats the remote exposes. In other systems we’ve seen things like CPU and IO wait on the remote but this could be anything up to something like presto operator summaries.

File: presto-spi/src/main/java/com/facebook/presto/spi/ConnectorPageSource.java [hunk 3]
```diff
@@ -28,6 +28,12 @@
      */
     long getCompletedBytes();
 
+    /**
+     * Gets the number of input rows processed by this page source so far.
+     * If number is not available, this method should return zero.
+     */
+    int getCompletedPositions();
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/ConnectorPageSource.java]: completed positions can go beyond Integer.MAX. Maybe use `long`.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftPageSource.java [hunk 4]
```diff
@@ -106,6 +107,12 @@ public long getCompletedBytes()
         return completedBytes;
     }
 
+    @Override
+    public int getCompletedPositions()
+    {
+        return 0;
```
Comment (highker) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftPageSource.java]: completedPositions

**Pull request merged**